### PR TITLE
Run OCI image package building and publishing concurrently

### LINKED ISSUE #378: Build and Publish plugins OCI images in parallel to improve the speed
**Describe the feature request**

As part of the builder plugin we are generating an OCI image package for the plugins and using that package to publish OCI images to the destination container registry.

We can us go routines to do things in parallel to improve the speed.

**Describe alternative(s) you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it

- As part of the builder plugin we are generating an OCI image package for the plugins and using that package to publish OCI images to the destination container registry.
- Before this change, the building and publishing of the OCI image were happening sequentially. This change uses go concurrency to run multiple things using go routines to speed up the building and publishing process.

**Performance comparison:**
- The OCI image build to build 7 plugins for 3 os-arch took `40.27s` compared to `164.53s` earlier which is approx **4** times faster.
- The OCI image publishing to publish 7 plugins for 3 os-arch took `21.64s` compared to `195.62s` earlier which is approx **9** times faster.
- The OCI image build and publishing combined of 7 plugins for 3 os-arch took `107.22s` compared to `360.87s` earlier which is approx **3.5** times faster.

### Which issue(s) this PR fixes

Fixes #378

### Describe testing done for PR

```
Building Plugin Packages: (make plugin-build-packages)

   Without Parallel Execution: 164.53 secs
   With Parallel Execution: 40.27 secs

Publishing Plugin Packages: (make plugin-plugin-packages) (excluding the building packages)
    
   Without Parallel Execution: 195.62 secs
   With Parallel Execution: 21.64 secs

Building & Publishing Plugin Packages Combined: (make plugin-plugin-packages) 

   Without Parallel Execution: 360.87 secs
   With Parallel Execution: 107.22 secs

```

### Release note

```release-note
Run OCI image package building and publishing concurrently when using `builder` plugin
```

### CODE REVIEW

**Comment by marckhouzam** [comment]
I was hoping to try this out including the change from #370 but there is a merge conflict.
@anujc25 will you be able to rebase this PR once #370 is merged?

File: cmd/plugin/builder/plugin/publish-packages.go
Comment (marckhouzam) [inline cmd/plugin/builder/plugin/publish-packages.go:55]: Same comment about an array probably not being thread-safe here

File: cmd/plugin/builder/plugin/publish-packages.go
Comment (marckhouzam) [inline cmd/plugin/builder/plugin/publish-packages.go:86]: Missing `return`

File: cmd/plugin/builder/plugin/build-packages.go
Comment (marckhouzam) [inline cmd/plugin/builder/plugin/build-packages.go:78]: This function call will be the first output shown to the user(`Validating Plugin Binary file: <path>`.
This output does not have the special thread ID, but it does show the path of the plugin binary, so the user can figure things out.

I suggest that we add a printout at the very beginning of each execution of the new `generatePluginPackage()` to indicate we are starting a new thread.

As for `helpers.ValidatePluginBinary()` I leave it up to you to decide if you want to include the threadId in the output or not.  Or even remove the printout completely (it is not that important).  Actually, I'm thinking removing it is maybe better?

File: cmd/plugin/builder/helpers/helper.go
Comment (marckhouzam) [inline cmd/plugin/builder/helpers/helper.go:112]: Why did we choose to use a different ID than for compilation?
Is it to get rid of non-ascii symbols?

File: cmd/plugin/builder/plugin/build-packages.go
Comment (anujc25) [reply cmd/plugin/builder/plugin/build-packages.go:62]: Good catch. I will update it to use channels. I didn't pay attention to these details. 👍

File: cmd/plugin/builder/plugin/build-packages.go
Comment (anujc25) [reply cmd/plugin/builder/plugin/build-packages.go]: Good point. Updated to pass threadID to the `ValidatePluginBinary` to make logs consistent.

File: cmd/plugin/builder/helpers/helper.go
Comment (anujc25) [reply cmd/plugin/builder/helpers/helper.go]: That was the initial thinking but I am updated to use the existing symbols instead of updating them in this PR. If needed we can update it in a separate PR.

**Review by marckhouzam — COMMENTED** [review]
Once last tweak about the concurrency and we're good.

File: cmd/plugin/builder/plugin/build-packages.go
Comment (chandrareddyp) [reply cmd/plugin/builder/plugin/build-packages.go:55]: @anujc25 Have you considered using a slice instead of a channel? I would like to know if there are compelling reasons to choose a channel over a slice.

File: cmd/plugin/builder/helpers/parallel.go
Comment (chandrareddyp) [inline cmd/plugin/builder/helpers/parallel.go:20]: any specific reason for reducing 2 from the number of cpu count?

File: cmd/plugin/builder/plugin/build-packages.go
Comment (anujc25) [reply cmd/plugin/builder/plugin/build-packages.go:55]: Thanks @marckhouzam for the details.

> It is a rare case to have more than one version, so I was wondering what was the risk if it actually happened AND there were more errors that we had allocated.

Considering, when we do `make plugin-build` the plugin manifest generated with that will always only have 1 version in it and there is no way to build the plugin with the multiple version in a single run. So it is indeed a rare case where the user updates the plugin manifest manually to have multiple versions.

File: cmd/plugin/builder/helpers/parallel.go
Comment (anujc25) [reply cmd/plugin/builder/helpers/parallel.go:20]: As mentioned in the comment above, this is to make sure we do not overwhelm the system by using all CPUs to build plugins and leave a few CPUs available to do other operations. This is a prior art where we build plugins in parallel and I am using the same functions to build plugin packages and publishing those in parallel.

**Review by marckhouzam — APPROVED** [review]
LGTM
Thanks for addressing all the comments @anujc25 !

**Pull request merged**
